src/gc.c

```diff
@@ -0,0 +1,225 @@
+
+  if (!term) {
+    // printf("No term for GC to inspect\n");
+    goto end;
+  }
+  RedisModule_Log(ctx, "debug", "Garbage collecting for term '%s'", term);
+  InvertedIndex *idx = Redis_OpenInvertedIndexEx(sctx, term, strlen(term), 1, &idxKey);
+  size_t totalRemoved = 0;
+  size_t totalCollected = 0;
+  if (idx) {
+    // printf("Garbage collecting for term %s\n", term);
+    int blockNum = 0;
+    int num = 100;
+
+    do {
+      size_t bytesCollected = 0;
+      size_t recordsRemoved = 0;
+
+      TimeSampler_Start(&ts);
+
+      blockNum = InvertedIndex_Repair(idx, &sctx->spec->docs, blockNum, num, &bytesCollected,
+                                      &recordsRemoved);
+      TimeSampler_End(&ts);
+      RedisModule_Log(ctx, "debug", "Repair took %lldns", TimeSampler_DurationNS(&ts));
+
+      /// update the statistics with the the number of records deleted
+      totalRemoved += recordsRemoved;
+      sctx->spec->stats.numRecords -= recordsRemoved;
+      sctx->spec->stats.invertedSize -= bytesCollected;
+      gc->stats.totalCollected += bytesCollected;
+      totalCollected += bytesCollected;
+
+      // blockNum 0 means error or we've finished
+      if (!blockNum) break;
+
+      // After each iteration we yield execution
+      // First we close the relevant keys we're touching
+      RedisModule_CloseKey(sctx->key);
+      RedisModule_CloseKey(idxKey);
+      SearchCtx_Free(sctx);
+
+      // now release the global lock
+      RedisModule_ThreadSafeContextUnlock(ctx);
+
+      // try to acquire it again...
+      RedisModule_ThreadSafeContextLock(ctx);
+
+      // reopen the context - it might have gone away!
+      sctx = NewSearchCtx(ctx, (RedisModuleString *)gc->keyName);
+      // sctx null --> means it was deleted and we need to stop right now
+      if (!sctx) break;
+      // reopen the inverted index - it might have gone away
+      idx = Redis_OpenInvertedIndexEx(sctx, term, strlen(term), 1, &idxKey);
+    } while (idx != NULL);
+  }
+
+  if (totalRemoved) {
+    RedisModule_Log(ctx, "notice", "Garbage collected %zd bytes in %zd records for term '%s'",
+                    totalCollected, totalRemoved, term);
+  }
+
+  free(term);
+  gc->stats.numCycles++;
+  gc->stats.effectiveCycles += totalRemoved > 0 ? 1 : 0;
+
+  // if we didn't remove anything - reduce the frequency a bit.
+  // if we did  - increase the frequency a bit
+
+  if (gc->timer) {  // the timer is NULL if we've been cancelled
+    if (totalRemoved > 0) {
+      gc->hz = MIN(gc->hz * 1.1, GC_MAX_HZ);
+    } else {
+      gc->hz = MAX(gc->hz * 0.99, GC_MIN_HZ);
+    }
+
+    RMUtilTimer_SetInterval(gc->timer, hzToTimeSpec(gc->hz));
+  }
+
+  RedisModule_Log(ctx, "debug", "New HZ: %f\n", gc->hz);
+
+end:
+  if (sctx) {
+    RedisModule_CloseKey(sctx->key);
+    SearchCtx_Free(sctx);
+  }
+  if (idxKey) RedisModule_CloseKey(idxKey);
+  RedisModule_ThreadSafeContextUnlock(ctx);
+}
+
+static void gc_onTerm(void *privdata) {
```

Comment:
Might help to have these 'contextless' functions in RMUtil... (i.e. for freeing memory - this way it might be able to have a static context and simply free it using that).